### PR TITLE
WIP: Merge plugins sdk into grafana code base

### LINKED ISSUE #19666: Move go plugin sdk code into Grafana Repository
**What would you like to be added**:

Move the https://github.com/grafana/grafana-plugin-sdk-go repo into this code base.

**Why is this needed**:

 - We need to investigate whether we can use a separate go module for the plugin sdk code
 - Need to decide if this is versioned with, our independently of Grafana's version.

To fill out: @marefr / @marcusolsson

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #19666

**Special notes for your reviewer**:

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -97,3 +96,7 @@ theOutput/
 
 # Ignore compilation stats from `yarn stats`
 compilation-stats.json
+
+# Ignore plugins sdk stuff
+./pkg/plugins/sdk/js/lib
+./pkg/plugins/sdk/node_modules
```
Comment (papagian) [inline .gitignore]: Not sure where the `node_modules ` while actually be located now

**Comment by aknuds1** [comment]
Should we fix linting issues in the SDK as part of this PR?

File: .gitignore (hunk 1 above)
Comment (aknuds1) [reply .gitignore]: Isn't there still a node_modules directory produced in the root? If you keep the old entry in .gitignore, it should apply to all levels of the tree.

Comment (aknuds1) [reply .gitignore]: Why are you changing the `node_modules` ignore pattern? The original pattern matches node_modules at the repo root for me, and it should also match at any other level of the tree. Unless I'm missing something, I think the original pattern should be kept.

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: pkg/plugins/sdk/dataframe/arrow_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"bytes"
+	"flag"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/dataframe"
```
Comment (kylebrandt) [inline pkg/plugins/sdk/dataframe/arrow_test.go]:
```suggestion
	"github.com/grafana/pkg/plugins/sdk/dataframe"
```

File: pkg/plugins/sdk/dataframe/dataframe_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
+package dataframe_test
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/dataframe"
```
Comment (kylebrandt) [inline pkg/plugins/sdk/dataframe/dataframe_test.go]:
```suggestion
	"github.com/grafana/pkg/plugins/sdk/dataframe"
```

File: pkg/plugins/sdk/go.mod [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module github.com/grafana/grafana-plugin-sdk-go
```
Comment (kylebrandt) [inline pkg/plugins/sdk/go.mod]:
```suggestion
module github.com/grafana/pkg/plugins/sdk
```

File: .gitignore (hunk 1 above)
Comment (papagian) [reply .gitignore]: sorry, there was a conflict while a was merging it and I got confused

Comment (aknuds1) [reply .gitignore]: @papagian OK :)

Comment (aknuds1) [reply .gitignore]: Just fix it back then and I'll approve I guess.

File: pkg/plugins/sdk/plugin.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package grafana
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/datasource"
```
Comment (aknuds1) [inline pkg/plugins/sdk/plugin.go]: I guess this has to be fixed.

File: pkg/plugins/sdk/grpc.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package grafana
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/datasource"
```
Comment (aknuds1) [inline pkg/plugins/sdk/grpc.go]: ☝️

File: pkg/plugins/sdk/go.mod [hunk 7]
```diff
@@ -0,0 +1,11 @@
+module github.com/grafana/pkg/plugins/sdk
+
+go 1.12
```
Comment (aknuds1) [inline pkg/plugins/sdk/go.mod]: Change to go 1.13 please.

File: pkg/plugins/sdk/example/datasource.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
+package main
+
+import (
+	"context"
+	"log"
+	"os"
+	"time"
+
+	gf "github.com/grafana/grafana-plugin-sdk-go"
+	"github.com/grafana/grafana-plugin-sdk-go/dataframe"
```
Comment (aknuds1) [inline pkg/plugins/sdk/example/datasource.go]: ☝️

File: pkg/plugins/sdk/datasource_test.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
+package grafana
+
+import (
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/dataframe"
```
Comment (aknuds1) [inline pkg/plugins/sdk/datasource_test.go]: ☝️

File: pkg/plugins/sdk/datasource.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
+package grafana
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/dataframe"
```
Comment (aknuds1) [inline pkg/plugins/sdk/datasource.go]: ☝️

File: pkg/plugins/sdk/datasource.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/datasource"
```
Comment (aknuds1) [inline pkg/plugins/sdk/datasource.go]: ☝️

**Comment by kylebrandt** [comment]
I don't think the tests in the sdk module are being run.  Probably due to https://github.com/golang/go/wiki/Modules#should-i-have-multiple-modules-in-a-single-repository 

>  `go test ./...` from the repository root will no longer test everything in the repository

**Comment by kylebrandt** [comment]
> go list -mod=vendor ./pkg/plugins/sdk/...

still fails with ` package github.com/grafana/grafana/pkg/plugins/sdk: cannot find package "."` unless I run it from the sdk dir.

**Comment by kylebrandt** [comment]
sdk module is tested by circle now. Need to look at makefile for testing this as well.

On an aside, Updated from go 1.12 to 1.13 and now it has a more proper error:

```
# kbrandt @ kbrandtlx in ~/go/github.com/grafana/grafana on git:19666-move-puligns-sdk x [11:56:06] C:1
$ go list -mod=vendor ./pkg/plugins/sdk/...
go: directory pkg/plugins/sdk is outside main module
go: directory pkg/plugins/sdk/dataframe is outside main module
go: directory pkg/plugins/sdk/example is outside main module
go: directory pkg/plugins/sdk/genproto/datasource is outside main module
```

**Comment by kylebrandt** [comment]
Also, I think to validate this (before merging) we probably need to use this module in some other library inside of the Grafana module to make sure the imports and vendoring are all sane from Grafana's perspective.

**Comment by papagian** [comment]
Great work @kylebrandt

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Looks good to me except for a couple of nits.

**Comment by papagian** [comment]
> Great work with this 👍 Besides Kyle's suggestion of verify that this can be used from Grafana main module we should also try to use this module in a local example app/library. @papagian or @aknuds1 are you currently looking into this?

I will try it in a local example and this will succeed also with gel-app.

File: pkg/plugins/sdk/datasource.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
+package grafana
```
Comment (aknuds1) [inline pkg/plugins/sdk/datasource.go]: I conferred with @marefr, we both think it makes more sense to name the package "sdk" to correspond to the containing directory. What do you think?

**Comment by papagian** [comment]
Re-opening this, it was closed because I tried to rename the branch.

**Pull request closed without merging**